In [1]:
# Initialize Otter
import otter
grader = otter.Notebook("hw02A.ipynb")

In [2]:
import numpy as np
import pandas as pd

import matplotlib
import matplotlib.pyplot as plt
import seaborn as sns
sns.set()
plt.style.use('fivethirtyeight')

from IPython.display import display, Image
def display_figure_for_grader(fig):
    plotly.io.write_image(fig, 'temp.png')
    display(Image('temp.png'))

### 0 导入数据

In [3]:
from pathlib import Path
data_dir = Path('.')
data_dir.mkdir(exist_ok=True)
file_path = data_dir / Path('data.zip')
dest_path = file_path

In [4]:
import zipfile
my_zip = zipfile.ZipFile(dest_path, 'r')
list_names = my_zip.namelist()
list_names

['data/',
 'data/bus.csv',
 'data/ins.csv',
 'data/ins2vio.csv',
 'data/vio.csv',
 'data/sf_zipcodes.json',
 'data/legend.csv']

In [5]:
my_zip = zipfile.ZipFile(dest_path, 'r')
for info in my_zip.infolist():
    print('{}\t{}'.format(info.filename, info.file_size))

data/	0
data/bus.csv	665365
data/ins.csv	1860919
data/ins2vio.csv	1032799
data/vio.csv	4213
data/sf_zipcodes.json	474
data/legend.csv	120


In [8]:
data_dir = Path('.')
my_zip.extractall(data_dir)


In [10]:
!dir {data_dir / Path("data")}

 驱动器 D 中的卷是 Data
 卷的序列号是 DA77-50B6

 D:\S2410W1036\xuexi\data100\fa24-student\hw\hw02A\data 的目录

2025-08-06  16:53    <DIR>          .
2025-08-06  16:53    <DIR>          ..
2025-08-06  16:53           665,365 bus.csv
2025-08-06  16:53         1,860,919 ins.csv
2025-08-06  16:53         1,032,799 ins2vio.csv
2025-08-06  16:53               120 legend.csv
2025-08-06  16:53               474 sf_zipcodes.json
2025-08-06  16:53             4,213 vio.csv
               6 个文件      3,563,890 字节
               2 个目录 39,888,257,024 可用字节


In [11]:
import os

def head(filename, lines=5):
    """
    Returns the first few lines of a file.

    filename: the name of the file to open
    lines: the number of lines to include

    return: A list of the first few lines from the file.
    """
    from itertools import islice
    with open(filename, "r") as f:
        return list(islice(f, lines))

data_dir = "./"
for f in list_names:
    if not os.path.isdir(f):
        print(head(data_dir + f, 5), "\n")

['"business id column","name","address","city","state","postal_code","latitude","longitude","phone_number"\n', '"1000","HEUNG YUEN RESTAURANT","3279 22nd St","San Francisco","CA","94110","37.755282","-122.420493","-9999"\n', '"100010","ILLY CAFFE SF_PIER 39","PIER 39  K-106-B","San Francisco","CA","94133","-9999","-9999","+14154827284"\n', '"100017","AMICI\'S EAST COAST PIZZERIA","475 06th St","San Francisco","CA","94103","-9999","-9999","+14155279839"\n', '"100026","LOCAL CATERING","1566 CARROLL AVE","San Francisco","CA","94124","-9999","-9999","+14155860315"\n'] 

['"iid","date","score","type"\n', '"100010_20190329","03/29/2019 12:00:00 AM","-1","New Construction"\n', '"100010_20190403","04/03/2019 12:00:00 AM","100","Routine - Unscheduled"\n', '"100017_20190417","04/17/2019 12:00:00 AM","-1","New Ownership"\n', '"100017_20190816","08/16/2019 12:00:00 AM","91","Routine - Unscheduled"\n'] 

['"iid","vid"\n', '"97975_20190725","103124"\n', '"85986_20161011","103114"\n', '"95754_2019032

In [12]:
# Path to the directory containing data
dsDir = Path('data')

bus = pd.read_csv(dsDir/'bus.csv', encoding='ISO-8859-1')
ins2vio = pd.read_csv(dsDir/'ins2vio.csv')
ins = pd.read_csv(dsDir/'ins.csv')
vio = pd.read_csv(dsDir/'vio.csv')

# This code is essential for the autograder to function properly. Do not edit
ins_test = ins

In [13]:
bus.head()

,business id column,name,address,city,state,postal_code,latitude,longitude,phone_number
0,1000,HEUNG YUEN RESTAURANT,3279 22nd St,San Francisco,CA,94110,37.755282,-122.420493,-9999
1,100010,ILLY CAFFE SF_PIER 39,PIER 39 K-106-B,San Francisco,CA,94133,-9999.000000,-9999.000000,14154827284
2,100017,AMICI'S EAST COAST PIZZERIA,475 06th St,San Francisco,CA,94103,-9999.000000,-9999.000000,14155279839
3,100026,LOCAL CATERING,1566 CARROLL AVE,San Francisco,CA,94124,-9999.000000,-9999.000000,14155860315
4,100030,OUI OUI! MACARON,2200 JERROLD AVE STE C,San Francisco,CA,94124,-9999.000000,-9999.000000,14159702675


In [15]:
bus.describe()

,business id column,latitude,longitude,phone_number
count,6253.000000,6253.000000,6253.000000,6.253000e+03
mean,60448.948984,-5575.337966,-5645.817699,4.701819e+09
std,36480.132445,4983.390142,4903.993683,6.667508e+09
min,19.000000,-9999.000000,-9999.000000,-9.999000e+03
25%,18399.000000,-9999.000000,-9999.000000,-9.999000e+03
50%,75685.000000,-9999.000000,-9999.000000,-9.999000e+03
75%,90886.000000,37.776494,-122.421553,1.415533e+10
max,102705.000000,37.824494,0.000000,1.415988e+10


In [16]:
assert all(bus.columns == ['business id column', 'name', 'address', 'city', 'state', 'postal_code',
                           'latitude', 'longitude', 'phone_number'])
assert 6250 <= len(bus) <= 6260

assert all(ins.columns == ['iid', 'date', 'score', 'type'])
assert 26660 <= len(ins) <= 26670

assert all(vio.columns == ['description', 'risk_category', 'vid'])
assert 60 <= len(vio) <= 65

assert all(ins2vio.columns == ['iid', 'vid'])
assert 40210 <= len(ins2vio) <= 40220

In [17]:
bus_summary = pd.DataFrame(**{'columns': ['business id column', 'latitude', 'longitude'],
 'data': {'business id column': {'50%': 75685.0, 'max': 102705.0, 'min': 19.0},
  'latitude': {'50%': -9999.0, 'max': 37.824494, 'min': -9999.0},
  'longitude': {'50%': -9999.0,
   'max': 0.0,
   'min': -9999.0}},
 'index': ['min', '50%', 'max']})

ins_summary = pd.DataFrame(**{'columns': ['score'],
 'data': {'score': {'50%': 76.0, 'max': 100.0, 'min': -1.0}},
 'index': ['min', '50%', 'max']})

vio_summary = pd.DataFrame(**{'columns': ['vid'],
 'data': {'vid': {'50%': 103135.0, 'max': 103177.0, 'min': 103102.0}},
 'index': ['min', '50%', 'max']})

from IPython.display import display

print('What we expect from your Businesses DataFrame:')
display(bus_summary)
print('What we expect from your Inspections DataFrame:')
display(ins_summary)
print('What we expect from your Violations DataFrame:')
display(vio_summary)

What we expect from your Businesses DataFrame:


,business id column,latitude,longitude
min,19.0,-9999.000000,-9999.0
50%,75685.0,-9999.000000,-9999.0
max,102705.0,37.824494,0.0


What we expect from your Inspections DataFrame:


,score
min,-1.0
50%,76.0
max,100.0


What we expect from your Violations DataFrame:


,vid
min,103102.0
50%,103135.0
max,103177.0


In [18]:
"""Run this cell to load this utility comparison function that we will use in various
tests below (both tests you can see and those we run internally for grading).

Do not modify the function in any way.
"""


def df_allclose(actual, desired, columns=None, rtol=5e-2):
    """Compare selected columns of two Dataframes on a few summary statistics.

    Compute the min, median and max of the two Dataframes on the given columns, and compare
    that they match numerically to the given relative tolerance.

    If they don't match, an AssertionError is raised (by `numpy.testing`).
    """
    # Summary statistics to compare on
    stats = ['min', '50%', 'max']

    # For the desired values, we can provide a full DF with the same structure as
    # the actual data, or pre-computed summary statistics.
    # We assume a pre-computed summary was provided if column is None. In that case,
    # `desired` *must* have the same structure as the actual's summary
    if columns is None:
        des = desired
        columns = desired.columns
    else:
        des = desired[columns].describe().loc[stats]

    # Extract summary stats from actual DF
    act = actual[columns].describe().loc[stats]

    return np.allclose(act, des, rtol)

### Q1 检查Business数据文件

In [19]:
bus.head()

,business id column,name,address,city,state,postal_code,latitude,longitude,phone_number
0,1000,HEUNG YUEN RESTAURANT,3279 22nd St,San Francisco,CA,94110,37.755282,-122.420493,-9999
1,100010,ILLY CAFFE SF_PIER 39,PIER 39 K-106-B,San Francisco,CA,94133,-9999.000000,-9999.000000,14154827284
2,100017,AMICI'S EAST COAST PIZZERIA,475 06th St,San Francisco,CA,94103,-9999.000000,-9999.000000,14155279839
3,100026,LOCAL CATERING,1566 CARROLL AVE,San Francisco,CA,94124,-9999.000000,-9999.000000,14155860315
4,100030,OUI OUI! MACARON,2200 JERROLD AVE STE C,San Francisco,CA,94124,-9999.000000,-9999.000000,14159702675


#### Q1a 把Business数据文件中的business id column列重命名为bid，检查bid唯一值的数量，是否每个值都是唯一的

In [20]:
bus = bus.rename(columns={'business id column': 'bid'})

In [23]:
is_bid_unique = len(bus['bid'].value_counts())
is_bid_unique

6253

In [24]:
grader.check("q1a")

q1a results: All test cases passed!

#### Q1b 找到最频繁的企业名称和最受欢迎的地址
先value_counts()，然后取前6个（index[:6]）<br>
这里要转成list或numpy的array<br>
测试的时候会用top_names[-1]去检查，如果仅仅取前六个+重置index的话，无法通过-1索引取得最后一个元素<br>
从而导致测试失败<br>

In [55]:
top_names = bus['name'].value_counts().head(6)
top_addresses = bus['address'].value_counts().head(6)

top_names = top_names.reset_index()['name'].tolist()

top_addresses = top_addresses.reset_index()['address'].tolist()

In [54]:
# top_names = bus['name'].value_counts().index[:6].tolist()
# top_addresses = bus['address'].value_counts().index[:6].tolist()
# #
# display(top_names)
# display(top_addresses)

["Peet's Coffee & Tea",
 'Starbucks Coffee',
 "McDonald's",
 'Jamba Juice',
 'STARBUCKS',
 'Proper Food']

['Off The Grid',
 '428 11th St',
 '2948 Folsom St',
 '3251 20th Ave',
 'Pier 41',
 '103 Horne Ave']

In [56]:
grader.check("q1b")

q1b results: All test cases passed!

#### Q1c 回答bus中一条记录代表什么
a 一个城市街区<br>
b 连锁餐厅<br>
c 餐厅的一个位置<br>

In [58]:
q1c = "C"

In [59]:
grader.check("q1c")

q1c results: All test cases passed!

### Q2 清理Business数据邮政编码：通过每条记录的邮政编码聚合城市各个地区的评级

#### Q2a 计算每个邮政编码有多少家餐厅

In [60]:
bus.head()

,bid,name,address,city,state,postal_code,latitude,longitude,phone_number
0,1000,HEUNG YUEN RESTAURANT,3279 22nd St,San Francisco,CA,94110,37.755282,-122.420493,-9999
1,100010,ILLY CAFFE SF_PIER 39,PIER 39 K-106-B,San Francisco,CA,94133,-9999.000000,-9999.000000,14154827284
2,100017,AMICI'S EAST COAST PIZZERIA,475 06th St,San Francisco,CA,94103,-9999.000000,-9999.000000,14155279839
3,100026,LOCAL CATERING,1566 CARROLL AVE,San Francisco,CA,94124,-9999.000000,-9999.000000,14155860315
4,100030,OUI OUI! MACARON,2200 JERROLD AVE STE C,San Francisco,CA,94124,-9999.000000,-9999.000000,14159702675


In [63]:
zip_counts = bus['postal_code'].value_counts()
print(zip_counts.to_string())

postal_code
94103         562
94110         555
94102         456
94107         408
94133         398
94109         382
94111         259
94122         255
94105         249
94118         231
94115         230
94108         229
94124         218
94114         200
-9999         194
94112         192
94117         189
94123         177
94121         157
94104         142
94132         132
94116          97
94158          90
94134          82
94127          67
94131          49
94130           8
94143           5
94301           2
94188           2
94101           2
CA              2
94013           2
941102019       1
941             1
95112           1
94105-2907      1
94102-5917      1
94124-1917      1
94621           1
95122           1
95132           1
95109           1
95133           1
95117           1
94901           1
94105-1420      1
94544           1
64110           1
94122-1909      1
00000           1
94080           1
Ca              1
94602           1
94129           

In [64]:
grader.check("q2a")

q2a results: All test cases passed!

#### Q2b 邮政编码中存在不合理的值如CA等<br>
第一步读取有效的旧金山邮政编码列表<br>

In [91]:
valid_zips = pd.read_json('data/sf_zipcodes.json')['zip_codes']

In [92]:
# valid_zips

0     94102
1     94103
2     94104
3     94105
4     94107
5     94108
6     94109
7     94110
8     94111
9     94112
10    94114
11    94115
12    94116
13    94117
14    94118
15    94119
16    94120
17    94121
18    94122
19    94123
20    94124
21    94125
22    94126
23    94127
24    94128
25    94129
26    94130
27    94131
28    94132
29    94133
30    94134
31    94137
32    94139
33    94140
34    94141
35    94142
36    94143
37    94144
38    94145
39    94146
40    94147
41    94151
42    94158
43    94159
44    94160
45    94161
46    94163
47    94164
48    94172
49    94177
50    94188
Name: zip_codes, dtype: int64

In [93]:
grader.check("q2bi")

q2bi results: All test cases passed!

In [94]:
valid_zips = valid_zips.astype("string")

In [95]:
type(valid_zips.dtype)

pandas.core.arrays.string_.StringDtype

第二步，构造一个只包含没有有效邮政编码的企业的“DataFrame”。

In [96]:
invalid_zip_bus = bus[~bus['postal_code'].isin(valid_zips)]
# invalid_zip_bus = ...
invalid_zip_bus.head(20)

,bid,name,address,city,state,postal_code,latitude,longitude,phone_number
22,100126,Lamas Peruvian Food Truck,Private Location,San Francisco,CA,-9999,-9999.000000,-9999.000000,-9999
68,100417,"COMPASS ONE, LLC",1 MARKET ST. FL,San Francisco,CA,94105-1420,-9999.000000,-9999.000000,14154324000
96,100660,TEAPENTER,1518 IRVING ST,San Francisco,CA,94122-1909,-9999.000000,-9999.000000,14155868318
109,100781,LE CAFE DU SOLEIL,200 FILLMORE ST,San Francisco,CA,94117-3504,-9999.000000,-9999.000000,14155614215
144,101084,Deli North 200,1 Warriors Way Level 300 North East,San Francisco,CA,94518,-9999.000000,-9999.000000,-9999
156,101129,Vendor Room 200,1 Warriors Way Level 300 South West,San Francisco,CA,-9999,-9999.000000,-9999.000000,-9999
177,101192,Cochinita #2,2 Marina Blvd Fort Mason,San Francisco,CA,-9999,-9999.000000,-9999.000000,14150429222
276,102014,"DROPBOX (Section 3, Floor 7)",1800 Owens St,San Francisco,CA,-9999,-9999.000000,-9999.000000,-9999
295,102245,Vessell CA Operations (#4),2351 Mission St,San Francisco,CA,-9999,-9999.000000,-9999.000000,-9999
298,10227,The Napper Tandy,3200 24th St,San Francisco,CA,-9999,37.752581,-122.416482,-9999


In [97]:
grader.check("q2bii")

q2bii results: All test cases passed!

#### Q2c

在上一个问题中，许多商家的邮政编码都是同一个无效值，这个值很可能用来表示“缺失的邮政编码”。这些商家是否都位于某个“有趣的地址”呢？为此，接下来的代码单元会统计每个 `address`（地址）下拥有这个特定缺失邮编的商家数量。


### 第一步
找出那个唯一的、常见的缺失邮政编码，并将其赋值给 `missing_postal_code`。然后，创建一个名为 `bus_missing` 的 DataFrame，只包含那些邮政编码等于 `missing_postal_code` 的商家记录。

**解释**：
这段话的目的是让你先找出数据中用来表示“邮编缺失”的那个特殊值，然后筛选出所有邮编为这个值的商家，方便后续分析这些商家是否有共同特征（比如都在同一个地址）。

In [102]:
missing_postal_code = invalid_zip_bus['postal_code'].value_counts().index[0]
bus_missing = bus[bus['postal_code'] == missing_postal_code]

In [104]:
grader.check("q2ci")

q2ci results: All test cases passed!

第二步，通过bus_missing查找每个地址（邮政编码都是无效的-9999）上的企业数量，得到missing_zip_address_count

In [110]:
missing_zip_address_count = bus_missing['address'].value_counts()
missing_zip_address_count.head(10)

address
Off The Grid                  39
Off the Grid                  10
OTG                            4
Approved Locations             3
Approved Private Locations     3
OFF THE GRID                   3
Justin Herman Plaza            2
Treasure Island                2
428 11th St                    2
Private Location               1
Name: count, dtype: int64

In [109]:
grader.check("q2cii")

q2cii results: All test cases passed!

### Q2e 对于具有无效邮政编码的记录，除了如Q2d中所说，是统一的无效编码-9999的餐厅，还有的编码无效是因为使用了完整的9位而不是5位<br>
对于这种记录，先新建postal5列，存储'postal_code'列的前五位数字<br>
然后对于不在valid_zips有效邮政编码列表中的postal5，将这一项设置为'None'

In [115]:
bus['postal5'] = None
bus['postal5'] = bus['postal_code'].str[:5]

bus.loc[~bus['postal5'].isin(valid_zips), 'postal5'] = None
# Checking the corrected postal5 column
bus.loc[invalid_zip_bus.index, ['bid', 'name', 'postal_code', 'postal5']]

,bid,name,postal_code,postal5
22,100126,Lamas Peruvian Food Truck,-9999,None
68,100417,"COMPASS ONE, LLC",94105-1420,94105
96,100660,TEAPENTER,94122-1909,94122
109,100781,LE CAFE DU SOLEIL,94117-3504,94117
144,101084,Deli North 200,94518,None
...,...,...,...,...
6173,99369,HOTEL BIRON,94102-5917,94102
6174,99376,Mashallah Halal Food truck Ind,-9999,None
6199,99536,FAITH SANDWICH #2,94105-2907,94105
6204,99681,Twister,95112,None


In [116]:
grader.check("q2e")

q2e results: All test cases passed!

### Q3 检查ins文件/DataFrame，

In [117]:
ins.head(5)

,iid,date,score,type
0,100010_20190329,03/29/2019 12:00:00 AM,-1,New Construction
1,100010_20190403,04/03/2019 12:00:00 AM,100,Routine - Unscheduled
2,100017_20190417,04/17/2019 12:00:00 AM,-1,New Ownership
3,100017_20190816,08/16/2019 12:00:00 AM,91,Routine - Unscheduled
4,100017_20190826,08/26/2019 12:00:00 AM,-1,Reinspection/Followup


Q3a 检查iid是否都是唯一值（类似Q1a）

In [119]:
is_ins_iid_unique = ins['iid'].is_unique
is_ins_iid_unique

True

In [120]:
grader.check("q3a")

q3a results: All test cases passed!

### Q3b 观察到iid似乎由bid-时间组成，新建一列bid，存储每条记录的bid

In [121]:
ins['bid'] = ins['iid'].str.split('_').str[0].astype(int)
ins.head(5)

,iid,date,score,type,bid
0,100010_20190329,03/29/2019 12:00:00 AM,-1,New Construction,100010
1,100010_20190403,04/03/2019 12:00:00 AM,100,Routine - Unscheduled,100010
2,100017_20190417,04/17/2019 12:00:00 AM,-1,New Ownership,100017
3,100017_20190816,08/16/2019 12:00:00 AM,91,Routine - Unscheduled,100017
4,100017_20190826,08/26/2019 12:00:00 AM,-1,Reinspection/Followup,100017


In [122]:
grader.check("q3b")

q3b results: All test cases passed!

### Q3c 检查date列
第一步：ins['date']列中每行的类型

In [123]:
ins_date_type = type(ins['date'][0])
ins_date_type

str

In [124]:
grader.check("q3ci")

q3ci results: All test cases passed!

第二步，将原本为字符串str的date列转换为时间戳TimeStamp类型，使用pd.to_datetime()函数创建新的ins['TimeStamp']列<br>

In [125]:
format_string = '%m/%d/%Y %I:%M:%S %p'
ins['timestamp'] = pd.to_datetime(ins['date'], format=format_string)
ins['timestamp']

0       2019-03-29
1       2019-04-03
2       2019-04-17
3       2019-08-16
4       2019-08-26
           ...    
26658   2018-09-24
26659   2018-11-02
26660   2019-09-09
26661   2017-12-07
26662   2018-08-08
Name: timestamp, Length: 26663, dtype: datetime64[ns]

In [126]:
grader.check("q3cii")

q3cii results: All test cases passed!

第三步，查看检查数据ins中最早和最晚的日期

In [127]:
earliest_date = min(ins['timestamp'])
latest_date = max(ins['timestamp'])
print("Earliest Date:", earliest_date)
print("Latest Date:", latest_date)

Earliest Date: 2016-10-04 00:00:00
Latest Date: 2019-11-28 00:00:00


In [128]:
grader.check("q3ciii")

q3ciii results: All test cases passed!

第四步，创建新列ins['year']，存储每条记录的年份，使用Series的dt.year属性<br>

In [129]:
ins['year'] = ins['timestamp'].dt.year

In [130]:
grader.check("q3civ")

q3civ results: All test cases passed!

### Q4 一些分析
最终目标：分别在所有的偶数邮政编码和奇数邮政编码中找出名字最长和有效电话号码最长的餐馆。<br>
Q4a 查看奇数和偶数的邮政编码的餐厅名称间是否存在差异<br>
新建'name_length'列，存储每条记录的餐厅名称长度<br>

In [131]:
bus.head()

,bid,name,address,city,state,postal_code,latitude,longitude,phone_number,postal5
0,1000,HEUNG YUEN RESTAURANT,3279 22nd St,San Francisco,CA,94110,37.755282,-122.420493,-9999,94110
1,100010,ILLY CAFFE SF_PIER 39,PIER 39 K-106-B,San Francisco,CA,94133,-9999.000000,-9999.000000,14154827284,94133
2,100017,AMICI'S EAST COAST PIZZERIA,475 06th St,San Francisco,CA,94103,-9999.000000,-9999.000000,14155279839,94103
3,100026,LOCAL CATERING,1566 CARROLL AVE,San Francisco,CA,94124,-9999.000000,-9999.000000,14155860315,94124
4,100030,OUI OUI! MACARON,2200 JERROLD AVE STE C,San Francisco,CA,94124,-9999.000000,-9999.000000,14159702675,94124


In [132]:
bus['name_length'] = bus['name'].str.len()

In [133]:
grader.check("q4a")

q4a results: All test cases passed!

Q4b 去除bus中邮政编码为None的部分，得到新的'bus_valid' DataFrame<br>
使用isna()函数<br>
isna()函数返回一个布尔值DataFrame，表示每个元素是否为NaN（缺失值）。也就是说，返回True时，说明是None，返回False时说明不是None<br>
所以下面要取反

In [139]:
bus_valid = bus[~bus['postal5'].isna()]
bus_valid.head(10)

,bid,name,address,city,state,postal_code,latitude,longitude,phone_number,postal5,name_length
0,1000,HEUNG YUEN RESTAURANT,3279 22nd St,San Francisco,CA,94110,37.755282,-122.420493,-9999,94110,21
1,100010,ILLY CAFFE SF_PIER 39,PIER 39 K-106-B,San Francisco,CA,94133,-9999.000000,-9999.000000,14154827284,94133,21
2,100017,AMICI'S EAST COAST PIZZERIA,475 06th St,San Francisco,CA,94103,-9999.000000,-9999.000000,14155279839,94103,27
3,100026,LOCAL CATERING,1566 CARROLL AVE,San Francisco,CA,94124,-9999.000000,-9999.000000,14155860315,94124,14
4,100030,OUI OUI! MACARON,2200 JERROLD AVE STE C,San Francisco,CA,94124,-9999.000000,-9999.000000,14159702675,94124,16
5,100036,Hula Truck (#2),2 Marina Blvd,San Francisco,CA,94123,-9999.000000,-9999.000000,-9999,94123,15
6,100039,GENKI CREPES & MINI MART,330 CLEMENT ST,San Francisco,CA,94118,-9999.000000,-9999.000000,14155376414,94118,24
7,100041,UNCLE LEE CAFE,3608 BALBOA ST,San Francisco,CA,94121,-9999.000000,-9999.000000,-9999,94121,14
8,100055,Twirl and Dip,335 Martin Luther King Jr. Dr,San Francisco,CA,94118,-9999.000000,-9999.000000,14155300260,94118,13
9,100058,SF PITA HUB,475 06TH ST,San Francisco,CA,94103,-9999.000000,-9999.000000,14155642006,94103,11


In [140]:
grader.check("q4b")

q4b results: All test cases passed!

Q4c 得到is_even对应postal5是否为偶数<br>

In [142]:
is_even = bus_valid['postal5'].astype(int) % 2 == 0

In [143]:
grader.check("q4c")

q4c results: All test cases passed!

Q4d 找到邮政编码为偶数中企业名称最短的企业

In [146]:
shortest_name_even = bus_valid[is_even].sort_values(by='name_length').iloc[0]['name']
shortest_name_even

'GAI'

In [147]:
grader.check("q4d")

q4d results: All test cases passed!

Q4e 考虑bus_valid中邮编为偶数且电话号码有效的餐厅<br>
无效电话号码类似于2b中的常见无效邮编<br>
检查在两个条件同时满足的情况下，最短的企业名称是否还是'GAI'

In [151]:
missing_phone_number = bus['phone_number'].value_counts().index[0]
bus_valid = bus_valid[bus_valid['phone_number'] != missing_phone_number]

In [152]:
same_shortest_name = bus_valid[is_even].sort_values(by='name_length').iloc[0]['name'] == 'GAI'
same_shortest_name

C:\Users\e\AppData\Local\Temp\ipykernel_10524\754509645.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  same_shortest_name = bus_valid[is_even].sort_values(by='name_length').iloc[0]['name'] == 'GAI'


False

In [153]:
grader.check("q4e")

q4e results: All test cases passed!